nifi-nar-bundles/nifi-hbase-bundle/nifi-hbase-processors/src/main/java/org/apache/nifi/hbase/PutHBaseRecord.java

```diff
@@ -0,0 +1,309 @@
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        final RecordReaderFactory recordParserFactory = context.getProperty(RECORD_READER_FACTORY)
+                .asControllerService(RecordReaderFactory.class);
+        List<PutFlowFile> flowFiles = new ArrayList<>();
+        final String tableName = context.getProperty(TABLE_NAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String rowFieldName = context.getProperty(ROW_FIELD_NAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String columnFamily = context.getProperty(COLUMN_FAMILY).evaluateAttributeExpressions(flowFile).getValue();
+        final String fieldEncodingStrategy = context.getProperty(FIELD_ENCODING_STRATEGY).getValue();
+        final long start = System.nanoTime();
+        int index = 0;
+        int columns = 0;
+        boolean failed = false;
+        String startIndexStr = flowFile.getAttribute("restart.index");
+        int startIndex = -1;
+        if (startIndexStr != null) {
+            startIndex = Integer.parseInt(startIndexStr);
+        }
+
+        PutFlowFile first = null;
+        PutFlowFile last  = null;
+        try (RecordReader reader = recordParserFactory.createRecordReader(flowFile, session.read(flowFile), getLogger())) {
+            Record record;
+            if (startIndex >= 0) {
+                while ( index++ < startIndex && (reader.nextRecord()) != null) {}
+            }
+
+            while ((record = reader.nextRecord()) != null) {
+                PutFlowFile putFlowFile = createPut(context, record, reader.getSchema(), flowFile, rowFieldName, columnFamily, fieldEncodingStrategy);
+                flowFiles.add(putFlowFile);
+                if (index == 0) {
+                    first = putFlowFile;
+                }
+                index++;
+
+                if (flowFiles.size() == batchSize) {
+                    columns += addBatch(tableName, flowFiles);
+                    last = flowFiles.get(flowFiles.size() - 1);
+                    flowFiles = new ArrayList<>();
+                }
+            }
+            if (flowFiles.size() > 0) {
+                columns += addBatch(tableName, flowFiles);
+                last = flowFiles.get(flowFiles.size() - 1);
+            }
+        } catch (Exception ex) {
+            getLogger().error("Failed to put records to HBase.", ex);
+            failed = true;
+        }
+        
+        if (!failed) {
+            long sendMillis = System.nanoTime() - start;
+            List<String> urls = Arrays.asList(getTransitUri(first), getTransitUri(last));
+            final String details = String.format("Put %d cells to HBase.", columns);
+            session.getProvenanceReporter().send(flowFile, urls.get(0), details, sendMillis);
```

Comment:
We could probably send a single event here since everything we sent came from the same flow file. 

We can override getTransitUri so that it doesn't include the row id since that only made sense before when it was a row per flow file:
```
protected String getTransitUri(PutFlowFile putFlowFile) {
        return "hbase://" + putFlowFile.getTableName();
    }
```
Then in the details we can have the number of records sent like you have, or possibly the range from start index to end index.

We should also consider sending a provenance event in the failure case. Maybe in the failure case we can look at the value of columns and if its > 0 then we know we sent some data successfully and we can report a send event in that case.